samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -172,49 +179,56 @@ class KafkaSystemProducer(systemName: String,
     if (dropProducerExceptions) {
       warn("Ignoring producer exception. All messages in the failed producer request will be dropped!")
 
-      if (isFatalException) {
-        producerLock.synchronized {
-          // Prevent each callback from recreating producer for the same failure.
-          if (currentProducer == producer) {
-            info("Creating a new producer for system %s." format systemName)
-            try {
-              currentProducer.close(0, TimeUnit.MILLISECONDS)
-            } catch {
-              case exception: Exception => error("Exception while closing producer.", exception)
-            }
-            producer = getProducer()
-          }
-        }
-      }
-    } else {
-      // If there is an exception in the callback, it means that the Kafka producer has exhausted the max-retries
-      // Close producer to ensure messages queued in-flight are not sent and hence, avoid re-ordering
-      // This works because there is only 1 callback thread and no sends can complete until the callback returns.
-      if (isFatalException) {
-        fatalException.compareAndSet(null, producerException)
+      // Prevent each callback from closing and nulling producer for the same failure.
+      if (currentProducer == producerRef.get()) {
```

Comment:
If both currentProducer and producerRef.get() are null here, it can have null pointer exception. I would recommend to have a null check to make the code more robust.

Reply:
1. The java doc specifies that currentProducer must not be null
2. If currentProducer was null, the right thing to do is throw an exception because we can't properly handle the exception, so then the debate is NPE or other exception. I don't see the value.